File: logica/real.cpp
#include <QRegExpValidator>
#include "real.h"

const QString Real::regExp = "^(\\+|-)?([0-9]+(\\.[0-9])?[0-9]*)$";

Real::Real(double initVal) {
    this->setR(initVal);
}

void Real::setR(double valToSet) {
    this->rVal = valToSet;
}

double Real::getR() const {
    return this->rVal;
}

void Real::string(QString str) {
    str = str.simplified().toLower();
    str = str.replace( " ", "" );

    // Inizializziamo un validator per verificare il pattern
    QRegExpValidator validator;
    QRegExp rexp(regExp);
    validator.setRegExp(rexp);
    int pos = 0;
    if (validator.validate(str, pos) != QRegExpValidator::State::Acceptable) {
        throw exce_kalk(str.prepend("Formato numero non corretto\n").toStdString());
        return;
    }

    setR(str.toDouble());
}

Real& Real::operator+(const Real& b) const {
    Real* res = new Real();
    res->rVal = this->rVal + b.rVal;
    return *res;
}
Real& Real::operator-(const Real& b) const {
    Real* res = new Real();
    res->rVal = this->rVal - b.rVal;
    return *res;
}
Real& Real::operator*(const Real& b) const {
    Real* res = new Real();
    res->rVal = this->rVal * b.rVal;
    return *res;
}
Real& Real::operator/(const Real& b) const {
    Real* res = new Real();
    if (b.rVal == 0)
        throw exce_kalk("Errore divisione per 0");
    res->rVal = this->rVal / b.rVal;
    return *res;
}

QString Real::getString(unsigned int precision) const {
    return QString::number(this->rVal, 'f', precision);
}

File: java/src/com/kalk/numeri/Real.java
package com.kalk.numeri;

public class Real {
    private Double rVal = 0.0;

    public Real() {
        setR(0.0);
    }

    public Real(Double initVal) {
        setR(initVal);
    }

    public void setR(Double valToSet) {
        rVal = valToSet;
    }

    public Double getR() {
        return this.rVal;
    }

    public void string (String str) {
        str = str.replaceAll("\\s+","");
        setR(Double.valueOf(str));
    }

    public Real somma(Real b) {
        Real res = new Real();
        res.setR(getR() + b.getR());
        return res;
    }
    public Real sottrai (Real b) {
        Real res = new Real();
        res.setR(getR() - b.getR());
        return res;
    }
    public Real moltiplica (Real b) {
        Real res = new Real();
        res.setR(getR() * b.getR());
        return res;
    }
    public Real dividi(Real b) throws RuntimeException {
        Real res = new Real();
        if (b.getR() == 0)
            throw new RuntimeException("Errore divisione per 0");
        res.setR(getR() / b.getR());
        return res;
    }

    public String getString() {
        return String.format("%.2f", getR());
    }
}